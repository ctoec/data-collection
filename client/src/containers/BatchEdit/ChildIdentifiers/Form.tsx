import React from 'react';
import {
  EditFormProps,
  doesChildIdFormHaveErrors,
} from '../../../components/Forms';
import { ChildIdentifiersForm as ChildIdForm } from '../../../components/Forms';

/**
 * Special componnet that enables batch editing a child identifiers
 * form for a child. Records with errors in their identifier fields
 * will appear in the SideNav of batch edit, and as those errors are
 * addressed, the form will disappear after saving.
 */
export const ChildIdentifiersForm: React.FC<EditFormProps> = ({
  child,
  afterSaveSuccess,
  setAlerts,
  showField,
}) => {
  return (
    <>
      {doesChildIdFormHaveErrors(child) && (
        <ChildIdForm
          child={child}
          afterSaveSuccess={afterSaveSuccess}
          setAlerts={setAlerts}
          showField={showField}
        />
      )}
    </>
  );
};
