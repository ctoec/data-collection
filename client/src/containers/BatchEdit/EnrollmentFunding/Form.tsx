import React from 'react';
import {
  RecordFormProps,
  doesEnrollmentFormHaveErrors,
  doesFundingFormHaveErrors,
  FundingForm,
  EnrollmentForm,
} from '../../../components/Forms';

/**
 * Special Enrollment / Funding form for batch edit,
 * which chains individual edit forms for all enrollments and fundings
 * with missing or invalid information. As the user completes updating
 * each invdividual funding or enrollment form, they will disappear as
 * this component is re-rendered and the objects themselves no longer
 * have any missing info.
 */
export const EnrollmentFundingForm: React.FC<RecordFormProps> = ({
  child,
  afterSaveSuccess,
  setAlerts,
  showFieldOrFieldset,
  topHeadingLevel,
}) => {
  if (!child?.enrollments?.length) {
    return (
      <EnrollmentForm
        child={child}
        afterSaveSuccess={afterSaveSuccess}
        setAlerts={setAlerts}
        topHeadingLevel={topHeadingLevel}
      />
    );
  }
  // will return true if enrollment has validation errors,
  // including if any fundings have validation errors
  const enrollmentsWithValidationErrors = child?.enrollments
    ?.filter(({ id }) => doesEnrollmentFormHaveErrors(child, id))
    ?.map((enrollment) => ({
      ...enrollment,
      fundings: enrollment.fundings?.filter((funding) =>
        doesFundingFormHaveErrors(child, enrollment.id, funding.id)
      ),
    }));

  return (
    <>
      {enrollmentsWithValidationErrors?.map((enrollment) => {
        return (
          <>
            {/* If enrollment has site, then display siteName for context (if no site, then site field will be displayed) */}
            {enrollment?.site && (
              <div className="text-bold font-body-md">
                Site: {enrollment?.site.siteName}{' '}
              </div>
            )}
            {enrollment?.entry && (
              <div className="text-bold font-body-md">
                Enrollment dates: {enrollment?.entry.format('MM/D/YYYY')}
                {enrollment?.exit &&
                  ` - ${enrollment?.exit?.format('MM/D/YYYY')}`}
              </div>
            )}
            {enrollment?.fundings?.map((funding) => (
              <>
                {/* If funding has fundingSpace, then display funding source for context */}
                {funding.fundingSpace && (
                  <div className="text-bold font-body-md">
                    Funding source: {funding.fundingSpace.source}
                  </div>
                )}
                <FundingForm
                  id={`batch-edit-funding-${funding.id}`}
                  fundingId={funding.id}
                  enrollmentId={enrollment.id}
                  child={child}
                  afterSaveSuccess={afterSaveSuccess}
                  setAlerts={setAlerts}
                  showFieldOrFieldset={showFieldOrFieldset}
                  topHeadingLevel={topHeadingLevel}
                />
              </>
            ))}
            {(!enrollment?.fundings?.length ||
              doesEnrollmentFormHaveErrors(child, enrollment?.id, {
                excludeFundings: true,
              })) && (
              <>
                <EnrollmentForm
                  id={`batch-edit-enrollment-${enrollment?.id}`}
                  enrollmentId={enrollment?.id}
                  child={child}
                  afterSaveSuccess={afterSaveSuccess}
                  setAlerts={setAlerts}
                  showFieldOrFieldset={showFieldOrFieldset}
                  topHeadingLevel={topHeadingLevel}
                />
              </>
            )}
          </>
        );
      })}
    </>
  );
};
