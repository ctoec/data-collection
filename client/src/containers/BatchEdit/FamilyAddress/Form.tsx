import React from 'react';
import {
  EditFormProps,
  doesFamilyAddressFormHaveErrors,
} from '../../../components/Forms';
import { FamilyAddressForm as AddressForm } from '../../../components/Forms';

/**
 * Special componnet that enables batch editing a family address form for
 * a child. Records with errors in their family's address field will
 * appear in the SideNav of batch edit, and as those errors are
 * addressed, the form will disappear after saving.
 */
export const FamilyAddressForm: React.FC<EditFormProps> = ({
  child,
  afterSaveSuccess,
  setAlerts,
  showField,
}) => {
  return (
    <>
      {doesFamilyAddressFormHaveErrors(child) && (
        <AddressForm
          child={child}
          afterSaveSuccess={afterSaveSuccess}
          setAlerts={setAlerts}
          showField={showField}
        />
      )}
    </>
  );
};
