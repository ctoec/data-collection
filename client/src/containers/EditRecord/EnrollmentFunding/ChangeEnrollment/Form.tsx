import React, { useContext, useState } from 'react';
import {
  Form,
  ExpandCard,
  Button,
  FormSubmitButton,
} from '@ctoec/component-library';
import { ChangeEnrollment } from '../../../../shared/payloads';
import { Enrollment, Child, FundingSource } from '../../../../shared/models';
import {
  SiteField,
  EnrollmentStartDateField,
  CareModelField,
  AgeGroupField,
  NewFundingField,
} from '../../../../components/Forms/Enrollment/Fields';
import { ReportingPeriodField } from '../../../../components/Forms/Enrollment/Funding/Fields';
import UserContext from '../../../../contexts/UserContext/UserContext';
import { apiPost } from '../../../../utils/api';
import AuthenticationContext from '../../../../contexts/AuthenticationContext/AuthenticationContext';

type ChangeEnrollmentFormProps = {
  afterSaveSuccess: () => void;
  afterSaveFailure: (err: any) => void;
  child: Child;
  currentEnrollment?: Enrollment;
};

export const ChangeEnrollmentForm: React.FC<ChangeEnrollmentFormProps> = ({
  afterSaveSuccess,
  afterSaveFailure,
  child,
  currentEnrollment,
}) => {
  const { user } = useContext(UserContext);
  const sites = (user?.sites || []).filter(
    (site) => site.organizationId === child.organization?.id
  );

  const { accessToken } = useContext(AuthenticationContext);

  const [loading, setLoading] = useState(false);
  const onSubmit = (updatedData: ChangeEnrollment) => {
    setLoading(true);
    apiPost(`children/${child.id}/change-enrollment`, updatedData, {
      accessToken,
      jsonParse: false,
    })
      .then(afterSaveSuccess)
      .catch(afterSaveFailure)
      .finally(() => setLoading(false));
  };

  const activeFunding = currentEnrollment?.fundings?.find(
    (f) => !f.lastReportingPeriod
  );

  if (activeFunding && !activeFunding?.fundingSpace) {
    return (
      <div>
        <p>
          Add missing enrollment and/or funding information before you can
          change {child.firstName}'s enrollment
        </p>
        <ExpandCard>
          <Button text="Cancel" appearance="outline" />
        </ExpandCard>
      </div>
    );
  }

  return (
    <Form<ChangeEnrollment>
      id="change-enrollment-form"
      className="usa-form"
      data={{ newEnrollment: {} as Enrollment }}
      onSubmit={onSubmit}
    >
      <SiteField<ChangeEnrollment>
        sites={sites}
        enrollmentAccessor={(data) => data.at('newEnrollment')}
      />
      <EnrollmentStartDateField<ChangeEnrollment>
        enrollmentAccessor={(data) => data.at('newEnrollment')}
      />
      <CareModelField<ChangeEnrollment>
        enrollmentAccessor={(data) => data.at('newEnrollment')}
      />
      <AgeGroupField<ChangeEnrollment>
        enrollmentAccessor={(data) => data.at('newEnrollment')}
      />
      <NewFundingField<ChangeEnrollment>
        fundingAccessor={(data) =>
          data.at('newEnrollment').at('fundings').at(0)
        }
        getEnrollment={(data) => data.at('newEnrollment').value}
        orgId={child.organization.id}
      />

      {!!currentEnrollment && activeFunding && (
        <>
          <h3>Previous enrollment</h3>
          <ReportingPeriodField<ChangeEnrollment>
            accessor={(data) =>
              data.at('oldEnrollment').at('funding').at('lastReportingPeriod')
            }
            fundingSource={activeFunding.fundingSpace?.source as FundingSource} // This is known to have a value (check on line 58)
            isLast={true}
            optional={true}
            label={`Last reporting period for current ${activeFunding.fundingSpace?.source} - ${activeFunding.fundingSpace?.time} funding`}
          />
        </>
      )}

      <ExpandCard>
        <Button text="Cancel" appearance="outline" />
      </ExpandCard>
      <FormSubmitButton
        text={loading ? 'Changing enrollment...' : 'Change enrollment'}
        disabled={loading}
      />
    </Form>
  );
};
