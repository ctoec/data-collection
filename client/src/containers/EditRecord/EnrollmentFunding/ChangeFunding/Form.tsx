import React, { useContext, useState } from 'react';
import { Form, Button, FormSubmitButton } from '@ctoec/component-library';
import { ChangeFunding } from '../../../../shared/payloads';
import { NewFundingField } from '../../../../components/Forms/Enrollment/Fields';
import { ReportingPeriodField } from '../../../../components/Forms/Enrollment/Funding/Fields';
import { Enrollment } from '../../../../shared/models';
import { apiPost } from '../../../../utils/api';
import AuthenticationContext from '../../../../contexts/AuthenticationContext/AuthenticationContext';

type ChangeFundingFormProps = {
  orgId: number;
  afterSaveSuccess: () => void;
  afterSaveFailure: (err: any) => void;
  changeType: 'start' | 'end';
  enrollment: Enrollment;
  hideForm: () => void;
};

export const ChangeFundingForm: React.FC<ChangeFundingFormProps> = ({
  orgId,
  afterSaveSuccess,
  afterSaveFailure,
  changeType,
  enrollment,
  hideForm,
}) => {
  const [loading, setLoading] = useState(false);
  const { accessToken } = useContext(AuthenticationContext);
  const onSubmit = (updatedData: ChangeFunding) => {
    setLoading(true);
    apiPost(`enrollments/${enrollment.id}/change-funding`, updatedData, {
      accessToken,
      jsonParse: false,
    })
      .then(afterSaveSuccess)
      .catch(afterSaveFailure)
      .finally(() => setLoading(false));
  };

  const activeFunding = enrollment.fundings?.find(
    (f) => !f.lastReportingPeriod
  );

  return (
    <Form<ChangeFunding>
      id="change-funding-form"
      className="usa-form"
      data={{}}
      onSubmit={onSubmit}
    >
      {changeType === 'start' && (
        <NewFundingField<ChangeFunding>
          fundingAccessor={(data) => data.at('newFunding')}
          getEnrollment={() => enrollment}
          orgId={orgId}
        />
      )}
      {!!activeFunding && (
        <ReportingPeriodField<ChangeFunding>
          accessor={(data) => data.at('oldFunding').at('lastReportingPeriod')}
          fundingSource={activeFunding.fundingSpace.source}
          isLast={true}
          label={`Last reporting period for current ${activeFunding.fundingSpace?.source} - ${activeFunding.fundingSpace?.time} funding`}
          optional={changeType === 'start'}
        />
      )}
      <Button text="Cancel" appearance="outline" onClick={hideForm} />
      <FormSubmitButton
        text={
          loading
            ? changeType === 'end'
              ? 'Ending funding...'
              : 'Changing funding...'
            : changeType === 'end'
            ? 'End current funding'
            : 'Change current funding'
        }
        disabled={loading}
      />
    </Form>
  );
};
