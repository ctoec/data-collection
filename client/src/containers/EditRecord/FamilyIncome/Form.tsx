import React, { useState } from 'react';
import { RecordFormProps } from '../../../components/Forms';
import {
  IncomeDetermination,
  UndefinableBoolean,
} from '../../../shared/models';
import { RedeterminationCard } from './RedeterminationCard';
import { Button } from '@ctoec/component-library';
import { EditDeterminationCard } from './EditDeterminationCard';
import { FosterIncomeNotRequiredAlert } from '../../../components/Forms/FamilyIncome/FosterIncomeNotRequiredAlert';
import { getNextHeadingLevel, Heading } from '../../../components/Heading';

/**
 * Component enabling user to edit the family income portion of a child
 * record. Renders RedeterminationCard and EditDeterminationCard which
 * enable the user to edit existing family income determinations or add
 * a new one.
 */
export const FamilyIncomeForm: React.FC<RecordFormProps> = ({
  child,
  afterSaveSuccess,
  setAlerts,
  topHeadingLevel,
}) => {
  if (!child?.family) {
    throw new Error('Family income form rendered without family');
  }

  const [showRedeterminationForm, setShowRedeterminationForm] = useState(false);
  const [currentIsNew, setCurrentIsNew] = useState(false);

  if (child?.foster === UndefinableBoolean.Yes) {
    return (
      <>
        <Heading level={topHeadingLevel}>Family income determination</Heading>
        <FosterIncomeNotRequiredAlert />
      </>
    );
  }

  const determinations: IncomeDetermination[] =
    child.family.incomeDeterminations || []; // assume they're sorted
  const currentDetermination: IncomeDetermination | undefined =
    determinations[0];
  const pastDeterminations: IncomeDetermination[] = determinations.slice(1);

  return (
    <>
      <Heading level={topHeadingLevel}>Family income determination</Heading>
      {showRedeterminationForm && (
        <RedeterminationCard
          child={child}
          afterSaveSuccess={() => {
            setShowRedeterminationForm(false);
            setCurrentIsNew(true);
            afterSaveSuccess();
          }}
          onCancel={() => setShowRedeterminationForm(false)}
          setAlerts={setAlerts}
          topHeadingLevel={getNextHeadingLevel(topHeadingLevel)}
        />
      )}
      <div className="margin-top-1">
        <div className="display-flex align-center">
          <Heading
            level={getNextHeadingLevel(topHeadingLevel)}
            className="font-sans-md margin-top-2 margin-bottom-2"
          >
            {currentDetermination
              ? 'Current income determination'
              : 'No income information on record'}
          </Heading>
          {!showRedeterminationForm && (
            <Button
              className="margin-left-1"
              text="Redetermine"
              appearance="unstyled"
              onClick={() => {
                setShowRedeterminationForm(true);
              }}
            />
          )}
        </div>
        {currentDetermination && (
          <EditDeterminationCard
            child={child}
            determinationId={currentDetermination.id}
            afterSaveSuccess={afterSaveSuccess}
            isCurrent={true}
            currentIsNew={currentIsNew}
            setAlerts={setAlerts}
            // This is nested under the current income det header
            topHeadingLevel={getNextHeadingLevel(topHeadingLevel, 2)}
          />
        )}

        {!!pastDeterminations.length && (
          <>
            <div className="margin-top-1">
              <Heading level={getNextHeadingLevel(topHeadingLevel)}>
                Past income determinations
              </Heading>
              {pastDeterminations.map((determination) => (
                <EditDeterminationCard
                  child={child}
                  determinationId={determination.id}
                  afterSaveSuccess={afterSaveSuccess}
                  setAlerts={setAlerts}
                  topHeadingLevel={getNextHeadingLevel(topHeadingLevel, 2)}
                />
              ))}
            </div>
          </>
        )}
      </div>
    </>
  );
};
