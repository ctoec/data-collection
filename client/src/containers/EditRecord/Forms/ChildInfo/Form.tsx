import React, { useContext, useState } from 'react';
import {
  SasidField,
  FirstNameField,
  MiddleNameField,
  LastNameField,
  SuffixField,
  DateOfBirthField,
  BirthCertificateFormFieldSet,
  RaceField,
  EthnicityField,
  GenderField,
  FosterCheckbox,
} from './Fields';
import { Form, FormSubmitButton } from '@ctoec/component-library';
import { EditFormProps } from '../types';
import AuthenticationContext from '../../../../contexts/AuthenticationContext/AuthenticationContext';
import { Child } from '../../../../shared/models';
import { apiPut } from '../../../../utils/api';
import useIsMounted from '../../../../hooks/useIsMounted';

// import { useFocusFirstError } from '../../../hooks/useFocusFirstError';

export const ChildInfoForm = ({ child, onSuccess }: EditFormProps) => {
  const { accessToken } = useContext(AuthenticationContext);
  const isMounted = useIsMounted();
  const [saving, setSaving] = useState(false);
  // TODO: HOW ARE WE HANDLING ERRORS?
  // const [error, setError] = useState(null);

  // Focus should automatically be on first error on page
  // useFocusFirstError([error]);

  if (!child) {
    throw new Error('Child info rendered without child');
  }

  const onFormSubmit = (_child: Child) => {
    setSaving(true);
    apiPut(`children/${child.id}`, _child, { accessToken })
      .then(() => onSuccess())
      .catch((err) => {
        console.log(err);
      })
      .finally(() => (isMounted() ? setSaving(false) : null));
  };

  return (
    <Form<Child>
      className="ChildInfoForm usa-form"
      data={child}
      onSubmit={onFormSubmit}
      noValidate
      autoComplete="off"
    >
      <div className="grid-row grid-gap">
        <div className="mobile-lg:grid-col-12">
          <SasidField />
        </div>
        <div className="mobile-lg:grid-col-9">
          <FirstNameField />
        </div>
        <div className="mobile-lg:grid-col-9">
          <MiddleNameField />
        </div>
        <div className="display-flex flex-row flex-align-end grid-row grid-gap">
          <div className="mobile-lg:grid-col-9">
            <LastNameField />
          </div>
          <div className="mobile-lg:grid-col-3">
            <SuffixField />
          </div>
        </div>
      </div>
      <h2>Date of birth</h2>
      <DateOfBirthField />
      <h2>Birth certificate</h2>
      <BirthCertificateFormFieldSet />
      <h2>Race</h2>
      <RaceField />
      <h2>Ethnicity</h2>
      <EthnicityField />
      <h2>Gender</h2>
      <GenderField />
      <h2>Foster</h2>
      <FosterCheckbox />
      <FormSubmitButton
        text={saving ? 'Saving...' : 'Save'}
        disabled={saving}
      />
    </Form>
  );
};
