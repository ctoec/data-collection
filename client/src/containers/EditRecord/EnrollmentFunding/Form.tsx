import React from 'react';
import { RecordFormProps } from '../../../components/Forms/types';
import { EditFundingCard } from './EditFundingCard';
import { ChangeEnrollmentCard } from './ChangeEnrollment/Card';
import { ChangeFundingCard } from './ChangeFunding/Card';
import { EditEnrollmentCard } from './EditEnrollmentCard';
import { CareModel, Enrollment, Funding } from '../../../shared/models';
import { getNextHeadingLevel, Heading } from '../../../components/Heading';
import { InlineIcon } from '@ctoec/component-library';

export const EnrollmentFundingForm: React.FC<RecordFormProps> = ({
  child,
  afterSaveSuccess,
  setAlerts,
  topHeadingLevel,
}) => {
  if (!child) {
    throw new Error('Enrollment funding form rendered without child');
  }

  // Separate enrollments into current (no end date) and past
  // (with end date). Either may not exist
  const enrollments: Enrollment[] = child.enrollments || [];
  const allEnrollmentsFunded = enrollments.every(
    (e) => e.fundings !== undefined && e.fundings.length > 0
  );

  // We create enrollments in the backend even if there's no data
  // (so that we can check validation errors), so this always has
  // positive length; need to check that the det has null fields
  const noRecordedEnrollments =
    enrollments.length === 1 &&
    enrollments[0].site === null &&
    enrollments[0].model === CareModel.Unknown &&
    enrollments[0].ageGroup === undefined &&
    enrollments[0].entry === undefined;
  const currentEnrollment: Enrollment | undefined = enrollments.find(
    (e) => !e.exit
  );
  const pastEnrollments: Enrollment[] = currentEnrollment
    ? enrollments.filter((e) => e.id !== currentEnrollment.id)
    : enrollments;

  const commonProps = {
    afterSaveSuccess,
    child,
    setAlerts,
  };

  /**
   * Simple way to check if a funding was created by the backend as
   * a placeholder for the card or actually contains real info to
   * display
   */
  const fundingHasNoInformation = (funding: Funding | undefined) => {
    if (!funding) return true;
    return (
      !funding.fundingSpace &&
      !funding.firstReportingPeriod &&
      !funding.lastReportingPeriod
    );
  };

  return (
    <>
      <Heading level={topHeadingLevel}>Enrollment and funding</Heading>
      {noRecordedEnrollments && (
        <p className="usa-error-message">
          <InlineIcon icon="attentionNeeded" /> At least one funded enrollment
          is required for each child record.
        </p>
      )}
      {!allEnrollmentsFunded && (
        <p className="usa-error-message">
          <InlineIcon icon="attentionNeeded" /> All enrollments must be
          associated with at least one funding.
        </p>
      )}
      {!noRecordedEnrollments && (
        <ChangeEnrollmentCard
          {...commonProps}
          topHeadingLevel={getNextHeadingLevel(topHeadingLevel)}
          currentEnrollment={currentEnrollment}
        />
      )}
      {currentEnrollment && (
        <>
          {!noRecordedEnrollments && (
            <Heading level={getNextHeadingLevel(topHeadingLevel)}>
              Current enrollment
            </Heading>
          )}
          <EditEnrollmentCard
            {...commonProps}
            key="edit-current-enrollment"
            isCurrent={true}
            enrollmentId={currentEnrollment.id}
            topHeadingLevel={getNextHeadingLevel(topHeadingLevel, 2)}
            noRecordedEnrollments={noRecordedEnrollments}
          />
          {currentEnrollment.fundings?.map(
            (funding) =>
              !noRecordedEnrollments && (
                <EditFundingCard
                  {...commonProps}
                  key={funding.id}
                  isCurrent={true}
                  fundingId={funding.id}
                  enrollmentId={currentEnrollment.id}
                  topHeadingLevel={getNextHeadingLevel(topHeadingLevel, 2)}
                />
              )
          )}
          {!noRecordedEnrollments &&
            !currentEnrollment.fundings?.some((f) =>
              fundingHasNoInformation(f)
            ) && (
              <ChangeFundingCard
                {...commonProps}
                enrollment={currentEnrollment}
                orgId={child.organization.id}
                // This heading should be nested under current enrollment
                topHeadingLevel={getNextHeadingLevel(topHeadingLevel, 2)}
              />
            )}
        </>
      )}
      {!!pastEnrollments.length && (
        <>
          <Heading level={getNextHeadingLevel(topHeadingLevel)}>
            Past enrollments
          </Heading>
          {pastEnrollments.map((enrollment) => (
            <>
              <EditEnrollmentCard
                {...commonProps}
                key={enrollment.id}
                enrollmentId={enrollment.id}
                topHeadingLevel={getNextHeadingLevel(topHeadingLevel, 2)}
              />
              {enrollment.fundings?.map((funding) => (
                <EditFundingCard
                  {...commonProps}
                  key={funding.id}
                  fundingId={funding.id}
                  enrollmentId={enrollment.id}
                  topHeadingLevel={getNextHeadingLevel(topHeadingLevel, 2)}
                />
              ))}
            </>
          ))}
        </>
      )}
    </>
  );
};
