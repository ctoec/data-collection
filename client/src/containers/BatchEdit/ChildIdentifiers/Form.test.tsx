import React from 'react';
import {
  snapshotTestHelper,
  accessibilityTestHelper,
  renderHelper,
} from '../../../testHelpers';
import { ChildIdentifiersForm } from './Form';
import {
  BirthCertificateType,
  Child,
  Family,
  Organization,
} from '../../../shared/models';

const child = {
  name: 'First Last',
  firstName: 'First',
  lastName: 'Last',
  birthCertificateType: BirthCertificateType.unavailable,
  id: '00000000-0000-0000-0000-000000000000',
  family: {} as Family,
  organization: { id: 1 } as Organization,
} as Child;

describe('BatchEdit', () => {
  describe('EnrollmentFundingForm', () => {
    snapshotTestHelper(
      <ChildIdentifiersForm
        child={child}
        afterSaveSuccess={jest.fn}
        setAlerts={jest.fn()}
        showField={jest.fn()}
      />
    );

    accessibilityTestHelper(
      <ChildIdentifiersForm
        child={child}
        afterSaveSuccess={jest.fn}
        setAlerts={jest.fn()}
        showField={jest.fn()}
      />
    );

    it('only shows forms for children with identifier validationErrors', async () => {
      const renderResult = await renderHelper(
        <ChildIdentifiersForm
          child={child}
          afterSaveSuccess={jest.fn}
          setAlerts={jest.fn()}
          showField={jest.fn()}
        />
      );

      // Assert there are only forms for objects with validation errors:
      // 1 enrollment and 1 funding.
      //
      // For now, we can only deduce presence of forms based on their submit
      // buttons (TODO: add some headings to the forms?)
      const formSubmitButtons = await renderResult.findAllByText('Save');
      expect(formSubmitButtons).toHaveLength(2);
    });
  });
});
