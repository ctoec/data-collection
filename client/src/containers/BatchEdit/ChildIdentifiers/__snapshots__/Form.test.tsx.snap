// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BatchEdit EnrollmentFundingForm matches snapshot 1`] = `<DocumentFragment />`;
